mu4/domain/notation/internal/notationplayback.cpp

```diff
@@ -0,0 +1,315 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2020 MuseScore BVBA and others
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#include "notationplayback.h"
+
+#include <cmath>
+
+#include "log.h"
+
+#include "libmscore/score.h"
+#include "libmscore/tempo.h"
+#include "libmscore/part.h"
+#include "libmscore/instrument.h"
+#include "libmscore/repeatlist.h"
+#include "libmscore/measure.h"
+#include "libmscore/segment.h"
+#include "libmscore/system.h"
+#include "libmscore/sym.h"
+#include "libmscore/page.h"
+#include "libmscore/staff.h"
+
+#include "audio/midi/event.h" //! TODO Remove me
+
+using namespace mu::domain::notation;
+using namespace mu::audio::midi;
+
+static EventType convertType(int type)
+{
+    switch (type) {
+    case Ms::ME_NOTEON:       return ME_NOTEON;
+    case Ms::ME_NOTEOFF:      return ME_NOTEOFF;
+    case Ms::ME_CONTROLLER:   return ME_CONTROLLER;
+    case Ms::ME_PITCHBEND:    return ME_PITCHBEND;
+    case Ms::ME_META:         return ME_META;
+
+    case Ms::ME_TICK1:        return ME_INVALID;
+    case Ms::ME_TICK2:        return ME_INVALID;
+    default: {
+        LOGE() << "unknown midi type: " << type;
+    }
+    }
+    return ME_INVALID;
+}
+
+NotationPlayback::NotationPlayback(IGetScore* getScore)
+    : m_getScore(getScore)
+{
+}
+
+std::shared_ptr<MidiStream> NotationPlayback::midiStream() const
+{
+    Ms::Score* score = m_getScore->score();
+    if (!score) {
+        return nullptr;
+    }
+
+    std::shared_ptr<MidiStream> stream = std::make_shared<MidiStream>();
+
+    makeInitData(stream->initData, score);
+
+    stream->request.onReceive(this, [stream](uint32_t tick) {
+        UNUSED(tick);
+        stream->stream.close();
```

Comment:
Because at the moment the transfer of parts of midi data is not implemented. At the moment, all data is given at once. When requesting new data, we close the channel to indicate that there will be no more data.

Reply:
Gotchya